### PR TITLE
Upgrade nginx reverse proxy

### PR DESCRIPTION
**Pull request by psavva**
Upgrade and use jwilder/docker-gen with a custom templates
Upgrade  LetsEncrypt  jrcs/letsencrypt-nginx-proxy-companion:2.2.0
Successfully tested this setup, and avoids http 502 Bad Gateway which we currently face setting up the docker images.

### CODE REVIEW

**Comment by psavva** [comment]
@miladsoft @sondreb 
Please review these changes.

More PRs coming for each of the chains.

**Comment by miladsoft** [comment]
Thank you very much for PR.
I checked this on two ubuntu 20.04 Linux servers and it works fine.

**Comment by psavva** [comment]
Added SBC service for now.
Once we have confirmation that all changes are acceptable, rest of chains to be modified and committed.

**Comment by psavva** [comment]
Added Bitcoin

**Comment by psavva** [comment]
City Chain updated.  Needs testing

**Comment by dangershony** [comment]
will we update all the chains before pushing?

**Comment by psavva** [comment]
Yes, all chains will be updated.
I'm looking for feedback if what I've done here is OK, and to proceed with the rest.

File: docker/BTC/docker-compose.yml [hunk 1]
```diff
@@ -23,39 +23,86 @@ services:
   indexer:
     container_name: btc-indexer
     image: blockcore/indexer:0.2.32
+    deploy:
+        resources:
+            limits:
+              cpus: 0.200
+              memory: 256M
```
Comment (sondreb) [inline docker/BTC/docker-compose.yml:30]: Have you tried running the Bitcoin indexer with only 256BM RAM? How long did the indexing take then?

File: docker/BTC/docker-compose.yml [hunk 2]
```diff
@@ -23,39 +23,86 @@ services:
...
     restart: unless-stopped
     depends_on:
-        - mongo
-        - chain
+      mongo:
+        condition: service_healthy
+      chain:
+        condition: service_healthy
+    networks:
+      - btcnetwork
```
Comment (sondreb) [inline docker/BTC/docker-compose.yml:47]: Does this remove the need for manually connecting networks after creation? That would be great =)

File: docker/SERVER/docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+version: '3.8'
```
Comment (sondreb) [inline docker/SERVER/docker-compose.yaml:1]: Why was this file named .yaml and old deleted?

**Review by sondreb — COMMENTED** [review]
Some feedback needed to continue

File: docker/SERVER/docker-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      - certs:/etc/nginx/certs:ro
+      - ./custom_settings.conf:/etc/nginx/conf.d/custom_settings.conf:ro
+    labels:
+      - "com.github.jrcs.letsencrypt_nginx_proxy_companion.nginx_proxy"
+    restart: always
+    networks: 
+      - commonnetwork
+      
+  docker-gen:
+    image: jwilder/docker-gen:0.8.4
```
Comment (sondreb) [inline docker/SERVER/docker-compose.yaml:22]: What is the reason for using nginx directly and docker-gen, as oppose to continue to rely on nginx-proxy?

File: docker/BTC/docker-compose.yml (hunk 1 above)
Comment (psavva) [reply docker/BTC/docker-compose.yml:30]: Please can you recommend resource limits and I'll implement for all chains.

I have tested SBC, and will coordinate with Milad to test all the chains, but would like to use what we have here as a baseline to make the changes for all chains.

File: docker/BTC/docker-compose.yml (hunk 2 above)
Comment (psavva) [reply docker/BTC/docker-compose.yml:47]: This is to specify which networks the service is able to access.

The section below allows you to specify networks to automatically create.
No need to create the network manually.

https://github.com/block-core/chaininfo/blob/57aea404812c4437245915dabfb0ddfe3c9aa052/docker/BTC/docker-compose.yml#L99

Comment (psavva) [reply docker/BTC/docker-compose.yml:47]: Yes. I need to manually connect to the network.

File: docker/SERVER/docker-compose.yaml (hunk 3 above)
Comment (psavva) [reply docker/SERVER/docker-compose.yaml:1]: I've replaced this whole file with a known working example I commonly use on my servers.

I can rename to yml as per the original, but basically it's a new file.

File: docker/SERVER/docker-compose.yaml (hunk 4 above)
Comment (psavva) [reply docker/SERVER/docker-compose.yaml:22]: It's a well tested setup I use on my servers.

Using docker-gen, you're able to separate and maintain any version of nginx at any time.  It's highly customizable using the template file.

If you don't like this approach, I can revert and get the network to be created automatically on nginx-proxy.

Please let me know what you prefer and I'll move forward with that.

**Comment by psavva** [comment]
@sondreb please also note that I'll update all chains prior to requesting this to be merged.

All chains will be tested, hopefully with the help of @miladsoft  prior to finalizing the PR

File: docker/SERVER/docker-compose.yaml (hunk 4 above)
Comment (sondreb) [reply docker/SERVER/docker-compose.yaml:22]: Would prefer the nginx-poxy image, it's kept up-to-date and less custom code needed for us? https://github.com/nginx-proxy/nginx-proxy

File: docker/SERVER/docker-compose.yaml (hunk 3 above)
Comment (sondreb) [reply docker/SERVER/docker-compose.yaml:1]: It's functionally the same file, even though all lines have changed. Better to keep the same extension as before =)

File: docker/BTC/docker-compose.yml (hunk 1 above)
Comment (sondreb) [reply docker/BTC/docker-compose.yml:30]: Just set it to something for the different chains, and I'll test and verify one by one when it's all merged and I need to deploy it.

File: docker/SERVER/docker-compose.yaml (hunk 4 above)
Comment (psavva) [reply docker/SERVER/docker-compose.yaml:22]: Ok, no problem.

File: docker/SERVER/docker-compose.yaml (hunk 3 above)
Comment (psavva) [reply docker/SERVER/docker-compose.yaml:1]: Ok

File: docker/BTC/docker-compose.yml (hunk 1 above)
Comment (psavva) [reply docker/BTC/docker-compose.yml:30]: Ok

**Comment by psavva** [comment]
Dear @sondreb 

Some Stats running BTC Node with 256 MB Ram:

```
CONTAINER ID   NAME          CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O   PIDS
9af7d3b5ff87   btc-indexer   3.00%     89.45MiB / 256MiB     34.94%    15.6MB / 18.1MB   0B / 0B     27  
0a6d7bd4c786   btc-chain     22.10%    986.1MiB / 5.859GiB   16.44%    329MB / 21.5MB    0B / 0B     33  
2404d0a7cbd1   btc-mongo     0.34%     432.4MiB / 4.883GiB   8.65%     10.5MB / 468kB    0B / 0B     40  
57e5f28044e0   letsencrypt   0.35%     20.64MiB / 24.96GiB   0.08%     1.36kB / 0B       0B / 0B     16  
4e0925085ca6   proxy         0.28%     84.06MiB / 24.96GiB   0.33%     190kB / 222kB     0B / 0B     34  
```

Looking at the above, we should probably restrict nginx proxy and the acme companion.
We can greatly reduce the memory of the mongodb, btc chain as well.

I suggest we put:
btc-indexer - keep at 256 MB
btc-chain  - 1024 MB
btc-mongo - 512 MB
letsencrypt - 128 MB
proxy - 256 MB

These look like sane enough numbers for an idle system.

Each deployment should assess their workload per server, and adjust their resource limits.
The reason I introduced resource limiting here is because when testing without limits, the server quickly becomes starved as full cpu/memory is available to all the containers, and with a small enough server, the system becomes unstable and slow.

I'm moving forward with the rest, and we can fine tune them as need be.

**Comment by psavva** [comment]
This is what it looks like with my above minimum memory recommendations:

```
CONTAINER ID   NAME          CPU %     MEM USAGE / LIMIT   MEM %     NET I/O           BLOCK I/O   PIDS
ff45a1556be7   letsencrypt   0.20%     18.38MiB / 128MiB   14.36%    796B / 0B         0B / 0B     13  
ecc59942472f   proxy         0.27%     84.69MiB / 256MiB   33.08%    16.5kB / 17kB     0B / 0B     34  
6c0c76e0e4f0   btc-chain     18.87%    497.5MiB / 1GiB     48.58%    35.5MB / 2.74MB   0B / 0B     38  
8e91bfadfbbe   btc-mongo     0.40%     369.5MiB / 512MiB   72.16%    19.1kB / 31.3kB   0B / 0B     36  
9af7d3b5ff87   btc-indexer   2.07%     135.8MiB / 256MiB   53.05%    50.5MB / 60.4MB   0B / 0B     27  
```

**Comment by psavva** [comment]
Notes on CITY Chain:
The "blockcore-bws-city" image is not published.
I'm guessing this is for the bitpay wallet.
Suggest that we comment out the "bws" related containers until we are in a position to publish them in dockerhub.

**Comment by psavva** [comment]
Update: got out of Memory Exception on Bitcoin Chain at 91879/67407 -24472, 
Updated to 2 GB, and letting it run.

**Comment by psavva** [comment]
Updating to 6GB - again out of memory when utilizing 2 GB on both BTC and CITY.

**Comment by psavva** [comment]
As an FYI, the exos-chain seems not to have any nodes running publically.

I've had it running for about 2 hours now, and no connections, and everytime I refresh the indexer status, the GetThresholdStateMetrics throws an "Object reference not set to an instance of an object".
Perhaps this can be handled more gracefully.

It happens when you call the function prior to any info about the blockchain being downloaded from the peers.

```
exos-chain         | fail: Blockcore.Features.RPC.RPCMiddleware[0]
exos-chain         |       Internal error while calling RPC Method
exos-chain         |       System.NullReferenceException: Object reference not set to an instance of an object.
exos-chain         |          at Blockcore.Base.Deployments.ThresholdConditionCache.GetThresholdStateMetrics(ChainedHeader indexPrev, ThresholdState[] thresholdStates)
exos-chain         |          at Blockcore.Features.RPC.Controllers.FullNodeController.GetBlockchainInfo()
exos-chain         |          at lambda_method151(Closure , Object , Object[] )
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, 
Object controller, Object[] arguments)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()
exos-chain         |       --- End of stack trace from previous location ---
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
exos-chain         |       --- End of stack trace from previous location ---
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope 
scope, Object state, Boolean isCompleted)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
exos-chain         |          at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
exos-chain         |          at Blockcore.Features.RPC.RPCMiddleware.InvokeSingleAsync(HttpContext httpContext, JObject requestObj)
exos-chain         | fail: Blockcore.Features.RPC.RPCMiddleware[0]
exos-chain         |       Internal error while calling RPC Method
exos-chain         |       System.NullReferenceException: Object reference not set to an instance of an object.
exos-chain         |          at Blockcore.Base.Deployments.ThresholdConditionCache.GetThresholdStateMetrics(ChainedHeader indexPrev, ThresholdState[] thresholdStates)
exos-chain         |          at Blockcore.Features.RPC.Controllers.FullNodeController.GetBlockchainInfo()
exos-chain         |          at lambda_method151(Closure , Object , Object[] )
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, 
Object controller, Object[] arguments)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()
exos-chain         |       --- End of stack trace from previous location ---
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
exos-chain         |       --- End of stack trace from previous location ---
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope 
scope, Object state, Boolean isCompleted)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
exos-chain         |          at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
exos-chain         |          at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
exos-chain         |          at Blockcore.Features.RPC.RPCMiddleware.InvokeSingleAsync(HttpContext httpContext, JObject requestObj)
```

**Comment by psavva** [comment]
Dear @sondreb

I have completed the changes as discussed for all the chains.
Please go ahead and review this PR, and let me know of any further feedback.

Best Regards
Panayiotis Savva

**Review by sondreb — APPROVED** [review]
Approving this change and doing verification for it.

**Comment by sondreb** [comment]
What version of docker and docker-compose is required to use the compose files after these changes? We need to get that added into the README, I'm unable to run them as they are right now.

`Unsupported config option for services.chain: 'deploy'`

The files are version "2.4", afaik "deploy" is part of version "3" of the compose schema.

Needs to be something like this for 2.4:

```
mem_limit: 6000m
mem_reservation: 100m
```

**Comment by sondreb** [comment]
All compose files have now been updated to use the correct resource limits for version "2.4".

**Pull request merged**